pymagicc/io/__init__.py

```diff
@@ -0,0 +1,613 @@
+        data: pd.DataFrame, pd.Series, np.ndarray or string
+            A pd.DataFrame or data file, or a numpy array of timeseries data if `columns` is specified.
+            If a string is passed, data will be attempted to be read from file.
+
+        columns: dict
+            Dictionary to use to write the metadata for each timeseries in data. MAGICCData will
+            also attempt to infer values from data. Any values in columns will be used in
+            preference to any values found in data. The default value for "model", "scenario"
+            and "climate_model" is "unspecified". See ``scmdata.ScmDataFrame`` for details.
+
+        kwargs:
+            Additional parameters passed to `pyam.core.read_files` to read non-standard files.
+        """
+        if not isinstance(data, str):
+            self.filepath = None
+            self.metadata = {}
+            super().__init__(data, columns=columns, **kwargs)
+        else:
+
+            filepath = data  # assume filepath
+            self.filepath = filepath
+            self.metadata, data, read_columns = _read_and_return_metadata_df(filepath)
+            columns = deepcopy(columns) if columns is not None else {}
+            for k, v in read_columns.items():
+                columns.setdefault(k, v)
+
+            columns.setdefault("model", ["unspecified"])
+            columns.setdefault("scenario", ["unspecified"])
+            columns.setdefault("climate_model", ["unspecified"])
+
+            super().__init__(data, columns=columns, **kwargs)
+
+    def _format_datetime_col(self):
+        time_srs = self["time"]
+        if isinstance(time_srs.iloc[0], datetime):
+            pass
+        elif isinstance(time_srs.iloc[0], int):
+            time_srs = [datetime(y, 1, 1) for y in to_int(time_srs)]
+        else:
+            time_srs = time_srs.apply(lambda x: convert_to_datetime(x))
+
+        self["time"] = time_srs
+
+    def _raise_not_loaded_error(self):
+        raise ValueError("File has not been read from disk yet")
+
+    @property
+    def is_loaded(self):
+        """bool: Whether the data has been loaded yet."""
+        return self._meta is not None
+
+    def append(self, other, inplace=False, constructor_kwargs={}, **kwargs):
+        """
+        Append any input which can be converted to MAGICCData to self.
+
+        Parameters
+        ----------
+        other : MAGICCData, pd.DataFrame, pd.Series, str
+            Source of data to append.
+
+        inplace : bool
+            If True, append ``other`` inplace, otherwise return a new ``MAGICCData``
+            instance.
+
+        constructor_kwargs : dict
+            Passed to ``MAGICCData`` constructor (only used if ``other`` is not a
+            ``MAGICCData`` instance).
+
+        **kwargs
+            Passed to ``super().append()``
+        """
+        if not isinstance(other, MAGICCData):
+            other = MAGICCData(other, **constructor_kwargs)
+
+        if inplace:
```

Comment:
this metadata handling can be fixed by ScmRun at some point (again not for this PR but maybe move into a new issue?)